probe-rs/src/architecture/arm/memory/adi_v5_memory_interface.rs

```diff
@@ -392,32 +462,98 @@ where
+        self.write_csw_register(access_port, csw)?;
+
+        let mut address = address;
+        self.write_tar_register(access_port, address)?;
+
+        // The maximum chunk size we can read before data overflows.
+        let max_chunk_size_bytes = 0x400;
+
+        let mut remaining_data_len = data.len();
+
+        let first_chunk_size_bytes = std::cmp::min(
+            max_chunk_size_bytes - (address as usize % max_chunk_size_bytes),
+            data.len(),
+        );
+
+        let mut data_offset = 0;
+
+        log::debug!(
+            "Read first block with len {} at address {:#08x}",
+            first_chunk_size_bytes,
+            address
+        );
+
+        let first_chunk_size_words = first_chunk_size_bytes;
+
+        self.read_ap_register_repeated(
+            access_port,
+            DRW { data: 0 },
+            &mut data_u32[data_offset..first_chunk_size_words],
+        )?;
+
+        remaining_data_len -= first_chunk_size_words;
+        address = address
+            .checked_add((first_chunk_size_words) as u64)
+            .ok_or(AccessPortError::OutOfBounds)?;
+        data_offset += first_chunk_size_words;
+
+        while remaining_data_len > 0 {
+            // The autoincrement is limited to the 10 lowest bits so we need to write the address
+            // every time it overflows.
+            self.write_tar_register(access_port, address)?;
 
-        // Read aligned block of 32-bit words
-        let mut buf32 = vec![0u32; aligned_len / 4];
-        self.read_32(access_port, aligned.start, &mut buf32)?;
+            let next_chunk_size_bytes = std::cmp::min(max_chunk_size_bytes, remaining_data_len);
+
+            log::debug!(
+                "Reading chunk with len {} at address {:#08x}",
+                next_chunk_size_bytes,
+                address
+            );
+
+            let next_chunk_size_words = next_chunk_size_bytes;
+
+            self.read_ap_register_repeated(
+                access_port,
+                DRW { data: 0 },
+                &mut data_u32[data_offset..(data_offset + next_chunk_size_words)],
+            )?;
 
-        // Convert 32-bit words to bytes
-        let mut buf8 = vec![0u8; aligned_len];
-        for (i, word) in buf32.into_iter().enumerate() {
-            buf8.pwrite_with(word, i * 4, LE).unwrap();
+            remaining_data_len -= next_chunk_size_words;
+            address
+                .checked_add((next_chunk_size_words) as u64)
+                .ok_or(AccessPortError::OutOfBounds)?;
+            data_offset += next_chunk_size_words;
         }
 
-        // Copy relevant part of aligned block to output data
-        let start = (address - aligned.start) as usize;
-        data.copy_from_slice(&buf8[start..start + data.len()]);
+        for (target, (i, source)) in data.iter_mut().zip(data_u32.iter().enumerate()) {
+            *target = ((*source >> (((start_address + i as u64) % 4) * 8)) & 0xFF) as u8;
```

Comment:
I don't understand what's going on here. Isn't the data in the source always in the lowest bits?

We read each u8 separately, so we get a u32 where only the lowest byte is used. Converting this into a u8 should just be a simple `u32 as u8`I think.

Reply:
Nope, unfortunately it is not. It's a bit weird. We have this shift somewhere else as well. And without it it does not pass unit tests. But this reminds me that I wanted to check the spec.

Reply:
I added a comment documenting this behavior.